traceapp/dashboard.go

```diff
@@ -0,0 +1,186 @@
+package traceapp
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"math"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"sourcegraph.com/sourcegraph/appdash"
+)
+
+// dasboardRow represents a single row in the dashboard. It is encoded to JSON.
+type dashboardRow struct {
+	Name                      string
+	Average, Min, Max, StdDev time.Duration
+	Timespans                 int
+	URL                       string
+}
+
+// newDashboardRow returns a new dashboard row with it's items calculated from
+// the given aggregation event (the returned row represents the whole
+// aggregation event).
+//
+// The returned row does not have the URL field set.
+func newDashboardRow(a appdash.AggregateEvent) dashboardRow {
+	row := dashboardRow{
+		Name:      a.Name,
+		Timespans: len(a.Times),
+	}
+
+	// Calculate sum and mean (row.Average), while determining min/max.
+	var sum time.Duration
+	for _, ts := range a.Times {
+		d := ts[1].Sub(ts[0])
+		sum += d
+		if row.Min == 0 || d < row.Min {
+			row.Min = d
+		}
+		if row.Max == 0 || d > row.Max {
+			row.Max = d
+		}
+	}
+	row.Average = sum / time.Duration(len(a.Times))
```

Comment:
Good eye, but yeah, we are guaranteed it's non-zero length. `newDashboardRow` is only called in one place, after filtering of the aggregation event occurs (ln 156-160):

```
			// Filter the event by our timeline.
			a, any := aggTimeFilter(a, start, end)
			if !any {
				continue
			}
```

Which only returns `any == true` if `len(a.Times) != 0` (ln 88):

```
	return cpy, len(cpy.Times) > 0
```